### PR TITLE
Improving wording related to static methods

### PR DESCRIPTION
**Pull request by moufmouf**
Following discussion in #16, I've tried to emphasize the fact that we expect factory methods to be stateless. I've also added a few words on what we mean by "stateless". Let me know if you think this is clear enough or if this still needs some work.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -74,6 +74,10 @@ If you know you will not be using the `$container` parameter or the `$getPreviou
     }
 ```
 
+A factory method returns the instance of the service it represents. A factory method does not need to ensure the unicity of the service returned. The container is in charge of calling the factory only once and then returning always the same service.
+
+A factory method should be **stateless**. The service created by the factory should only depend on the input parameters of the factory (`$container` and `$getPrevious`). If the factory needs to fetch parameters, those should be fetched from the container directly (see below).
```
Comment (mnapoli) [inline README.md]: :+1: really nice wording, I didn't think of that part

File: README.md [hunk 2]
```diff
@@ -74,6 +74,10 @@ If you know you will not be using the `$container` parameter or the `$getPreviou
     }
 ```
 
+A factory method returns the instance of the service it represents. A factory method does not need to ensure the unicity of the service returned. The container is in charge of calling the factory only once and then returning always the same service.
```
Comment (mnapoli) [inline README.md]: I find it too complicated honestly (and too much about objects/services). Here is a try:

> Factory methods are only responsible for returning the value for the container entry. Nothing should be cached by service providers, this is the responsibility of the container.

Still not satisfied though…

File: README.md [hunk 3]
```diff
@@ -218,7 +222,9 @@ If you register the service providers in the correct order in your container (A
 
 It comes from the assumption that container configuration is stateless, i.e. it doesn't depend on anything.
 
-If container configuration is not stateless, it introduces problems for compiled containers (e.g. Symfony) or containers that perform caching (e.g. PHP-DI). Using static methods somewhat guarantees everything to be stateless, which cannot be guaranteed otherwise.
+If container configuration is not stateless, it introduces problems for compiled containers (e.g. Symfony) or containers that perform caching (e.g. PHP-DI). Indeed, compiled containers and cached containers will call the `getServices` method in a process and the factories in another process. Relying on some state of the service provider class might therefore be a bad idea.
+
+Using static methods *conveys* the idea that everything should be stateless. Furthermore, since factories are supposed to be stateless, there is no need for a factory to rely on the service provider stat (using `$this`), hence the idea to make factories static.
```
Comment (mnapoli) [inline README.md]: > there is no need for a factory to rely on the service provider stat (using `$this`) […]

to:

> there is no need for a factory to rely on the service provider state (using `$this`) […]

File: README.md (hunk 2 above)
Comment (moufmouf) [inline README.md]: Your proposal is indeed much better. Let me try to improve a bit:

> Each factory method is responsible for returning a given entry of the container. Nothing should be cached by service providers, this is the responsibility of the container.

Good enough?

Comment (mnapoli) [inline README.md]: :+1:

**Comment by moufmouf** [comment]
Ok, we seem to agree on the new wording. I'm merging this.

**Pull request merged**